aptos-move/block-executor/src/captured_reads.rs

```diff
@@ -583,6 +615,150 @@ impl<T: Transaction> CapturedReads<T> {
+        module_name: &IdentStr,
+    ) -> CacheRead<Option<usize>> {
+        self.fetch_from_module_read(address, module_name, |v| v.map(|e| e.size_in_bytes()))
+    }
+
+    /// If the module has been previously read, returns its metadata.
+    pub(crate) fn fetch_module_metadata(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> CacheRead<Option<Vec<Metadata>>> {
+        self.fetch_from_module_read(address, module_name, |v| v.map(|e| e.metadata().to_vec()))
+    }
+
+    /// If the module has been previously read, returns its deserialized representation.
+    pub(crate) fn fetch_deserialized_module(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> CacheRead<Option<Arc<CompiledModule>>> {
+        self.fetch_from_module_read(address, module_name, |v| {
+            v.map(|e| e.compiled_module().clone())
+        })
+    }
+
+    /// If the module has been previously read, and has been verified, returns it. The important
+    /// difference from other APIs is that existing non-verified entries are treated as a cache
+    /// miss.
+    pub(crate) fn fetch_verified_module(
+        &self,
+        address: &AccountAddress,
+        module_name: &IdentStr,
+    ) -> CacheRead<VMResult<Option<Arc<Module>>>> {
+        match self.module_reads.get(&(address, module_name)) {
+            Some(ModuleRead::BlockCache(read)) => match read {
+                Some(v) => {
+                    if v.is_verified() {
+                        CacheRead::Hit(v.verified_module().map(|v| Some(v.clone())))
+                    } else {
+                        CacheRead::Miss
+                    }
+                },
+                None => CacheRead::Hit(Ok(None)),
+            },
+            None => CacheRead::Miss,
+            Some(ModuleRead::GlobalCache) => unreachable!("Global cache reads are not cached"),
+        }
+    }
+
+    /// For every module read that was captured, checks if the reads are still the same:
+    ///   1. Entries read from cross-block module cache are still valid.
+    ///   2. Entries that were not in block cache before are still not there.
+    ///   3. Entries that were in block cache have the same commit index.
+    pub(crate) fn validate_module_reads(
+        &self,
+        code_cache: &SyncCodeCache<
+            ModuleId,
+            Arc<MaybeCommitted<ModuleCacheEntry>>,
+            [u8; 32],
+            CachedScript,
+            VMError,
+        >,
+    ) -> bool {
+        if self.non_delayed_field_speculative_failure {
+            return false;
+        }
+
+        self.module_reads.iter().all(|(id, read)| match read {
+            ModuleRead::GlobalCache => CrossBlockModuleCache::is_valid(id),
+            ModuleRead::BlockCache(previous) => match previous {
+                Some(previous) => code_cache
+                    .module_cache()
+                    .contains_and(id, |m| m.commit_idx() == previous.commit_idx()),
```

Comment:
if this is the only place we use contains_and, we can instead have a map, or even better to not pass a closure something specifically that returns Option<commit index> and directly compare it to previous (instead of match)